File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/PrefHandlerImpl.java
package org.totschnig.myexpenses.preference;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import org.totschnig.myexpenses.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;

public class PrefHandlerImpl implements PrefHandler {
  private final Application context;
  private final SharedPreferences sharedPreferences;

  public PrefHandlerImpl(Application context, SharedPreferences sharedPreferences) {
    this.context = context;
    this.sharedPreferences = sharedPreferences;
  }

  @NonNull
  @Override
  public String getKey(PrefKey key) {
    return key.resId == 0 ? key.key : context.getString(key.resId);
  }

  @Override
  @Nullable
  public String getString(PrefKey key, String defValue) {
    return getString(getKey(key), defValue);
  }

  @Override
  @Nullable
  public String getString(String key, String defValue) {
    return sharedPreferences.getString(key, defValue);
  }

  @Override
  public void putString(PrefKey key, String value) {
    putString(getKey(key), value);
  }

  @Override
  public void putString(String key, String value) {
    sharedPreferences.edit().putString(key, value).apply();
  }

  @Override
  public boolean getBoolean(PrefKey key, boolean defValue) {
    return getBoolean(getKey(key), defValue);
  }

  @Override
  public boolean getBoolean(String key, boolean defValue) {
    return sharedPreferences.getBoolean(key, defValue);
  }

  @Override
  public void putBoolean(PrefKey key, boolean value) {
    putBoolean(getKey(key), value);
  }

  @Override
  public void putBoolean(String key, boolean value) {
    sharedPreferences.edit().putBoolean(key, value).apply();
  }

  @Override
  public int getInt(PrefKey key, int defValue) {
    return getInt(getKey(key), defValue);
  }

  @Override
  public int getInt(String key, int defValue) {
    return sharedPreferences.getInt(key, defValue);
  }

  @Override
  public void putInt(PrefKey key, int value) {
    putInt(getKey(key), value);
  }

  @Override
  public void putInt(String key, int value) {
    sharedPreferences.edit().putInt(key, value).apply();
  }

  @Override
  public long getLong(PrefKey key, long defValue) {
    return getLong(getKey(key), defValue);
  }

  @Override
  public long getLong(String key, long defValue) {
    return sharedPreferences.getLong(key, defValue);
  }

  @Override
  public void putLong(PrefKey key, long value) {
    putLong(getKey(key), value);
  }

  @Override
  public void putLong(String key, long value) {
    sharedPreferences.edit().putLong(key, value).apply();
  }

  @Override
  public void remove(PrefKey key) {
    remove(getKey(key));
  }

  @Override
  public void remove(String key) {
    sharedPreferences.edit().remove(key).apply();
  }

  @Override
  public boolean isSet(PrefKey key) {
    return isSet(getKey(key));
  }

  @Override
  public boolean isSet(String key) {
    return sharedPreferences.contains(key);
  }

  @Override
  public boolean matches(@NonNull String key, PrefKey... prefKeys) {
    for (PrefKey prefKey: prefKeys) {
      if (key.equals(getKey(prefKey)))
        return true;
    }
    return false;
  }

  @Override
  public void setDefaultValues(Context context) {
    PreferenceManager.setDefaultValues(context, R.xml.preferences, false);
  }

  @Override
  public void preparePreferenceFragment(PreferenceFragmentCompat preferenceFragmentCompat) {
    //NOOP overriden in test
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/PrefHandlerImpl.kt
package org.totschnig.myexpenses.preference

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import org.totschnig.myexpenses.R
import androidx.core.content.edit

open class PrefHandlerImpl(
    private val context: Application,
    private val sharedPreferences: SharedPreferences,
) : PrefHandler {
    override fun getKey(key: PrefKey) =
        if (key.resId == 0) key._key!! else context.getString(key.resId)

    override fun getString(key: String, defValue: String?) =
        sharedPreferences.getString(key, defValue)
    override fun putString(key: String, value: String?) {
        sharedPreferences.edit { putString(key, value) }
    }

    override fun getBoolean(key: String, defValue: Boolean) =
        sharedPreferences.getBoolean(key, defValue)
    override fun putBoolean(key: String, value: Boolean) {
        sharedPreferences.edit { putBoolean(key, value) }
    }

    override fun getInt(key: String, defValue: Int) = sharedPreferences.getInt(key, defValue)
    override fun putInt(key: String, value: Int) {
        sharedPreferences.edit { putInt(key, value) }
    }

    override fun getLong(key: String, defValue: Long) =
        sharedPreferences.getLong(key, defValue)
    override fun putLong(key: String, value: Long) {
        sharedPreferences.edit { putLong(key, value) }
    }

    override fun getFloat(key: String, defValue: Float) =
        sharedPreferences.getFloat(key, defValue)

    override fun getOrderedStringSet(key: String, separator: Char) =
        sharedPreferences.getString(key, null)?.let {
            LinkedHashSet(it.split(separator))
        }

    override fun putOrderedStringSet(key: String, value: Set<String>, separator: Char) {
        require(value.none { it.contains(separator) }) { "Cannot marshall set if any value contains '$separator'" }
        sharedPreferences.edit { putString(key, value.joinToString(separator.toString())) }
    }

    override fun getStringSet(key: String): Set<String>? =  sharedPreferences.getStringSet(key, null)
    override fun putStringSet(key: String, value: Set<String>) {
        sharedPreferences.edit { putStringSet(key, value) }
    }

    override fun remove(key: String) {
        sharedPreferences.edit { remove(key) }
    }

    override fun isSet(key: String) = sharedPreferences.contains(key)

    override fun matches(key: String, vararg prefKeys: PrefKey) =
        prefKeys.any { key == getKey(it) }

    override fun setDefaultValues(context: Context) {
        setDefaultValues(context, R.xml.preferences_advanced)
        setDefaultValues(context, R.xml.preferences_attach_picture)
        setDefaultValues(context, R.xml.preferences_backup_restore)
        setDefaultValues(context, R.xml.preferences_data)
        setDefaultValues(context, R.xml.preferences_feedback)
        setDefaultValues(context, R.xml.preferences_include_share)
        setDefaultValues(context, R.xml.preferences_io)
        setDefaultValues(context, R.xml.preferences_ocr)
        setDefaultValues(context, R.xml.preferences_print)
        setDefaultValues(context, R.xml.preferences_protection)
        setDefaultValues(context, R.xml.preferences_sync)
        setDefaultValues(context, R.xml.preferences_ui)
        setDefaultValues(context, R.xml.preferences_web_ui)
    }

    open fun setDefaultValues(context: Context, resId: Int) {
        PreferenceManager.setDefaultValues(context, resId, true)
    }

    override fun preparePreferenceFragment(preferenceFragmentCompat: PreferenceFragmentCompat) {
        //NOOP overridden in test
    }
}